sed/loader/flash/loader.py

```diff
@@ -0,0 +1,615 @@
+        for key in self.available_channels:
+            if (
+                self._config["dataframe"]["channels"][key]["format"]
+                == "per_pulse"
+            ):
+                if key == "dldAux":
+                    for aux_key in self._config["dataframe"]["channels"][key][
+                        "dldAuxChannels"
+                    ].keys():
+                        channels_per_pulse.append(aux_key)
+                else:
+                    channels_per_pulse.append(key)
+        return channels_per_pulse
+
+    @property
+    def channels_per_electron(self) -> List:
+        """Returns a list of channels with per_electron format"""
+        return [
+            key
+            for key in self.available_channels
+            if self._config["dataframe"]["channels"][key]["format"]
+            == "per_electron"
+        ]
+
+    @property
+    def channels_per_train(self) -> List:
+        """Returns a list of channels with per_train format"""
+        return [
+            key
+            for key in self.available_channels
+            if self._config["dataframe"]["channels"][key]["format"]
+            == "per_train"
+        ]
+
+    def reset_multi_index(self) -> None:
+        """Resets the index per pulse and electron"""
+        self.index_per_electron = None
+        self.index_per_pulse = None
+
+    def create_multi_index_per_electron(self, h5_file: h5py.File) -> None:
+        """Creates an index per electron using pulseId
+        for usage with the electron resolved pandas dataframe"""
+
+        # Macrobunch IDs obtained from the pulseId channel
+        [train_id, np_array] = self.create_numpy_array_per_channel(
+            h5_file,
+            "pulseId",
+        )
+
+        # Create a series with the macrobunches as index and
+        # microbunches as values
+        macrobunches = (
+            Series(
+                (np_array[i] for i in train_id.index),
+                name="pulseId",
+                index=train_id,
+            )
+            - self._config["dataframe"]["ubid_offset"]
+        )
+
+        # Explode dataframe to get all microbunch vales per macrobunch,
+        # remove NaN values and convert to type int
+        microbunches = macrobunches.explode().dropna().astype(int)
+
+        # Create temporary index values
+        index_temp = MultiIndex.from_arrays(
+            (microbunches.index, microbunches.values),
+            names=["trainId", "pulseId"],
+        )
+
+        # Calculate the electron counts per pulseId
+        # unique preserves the order of appearance
+        electron_counts = index_temp.value_counts()[index_temp.unique()].values
+
+        # Series object for indexing with electrons
+        electrons = (
+            Series(
+                [
+                    np.arange(electron_counts[i])
+                    for i in range(electron_counts.size)
+                ],
+            )
+            .explode()
+            .astype(int)
+        )
```

Comment:
this is so ugly... I know it comes from our linting, but we should maybe reconsider it, as a new line for each .something feels weird

Reply:
This comes from using black with the precommit hooks. I'd propose to use autopep8 or something then, as it doesn't format documents as such. Could you create an issue regarding this?